sw/device/lib/crypto/drivers/kmac.c

```diff
@@ -617,22 +614,20 @@ static status_t kmac_process_msg_blocks(kmac_operation_t operation,
     // NIST SP 800-185 (1 byte for encoding header and 255 byte for the encoded
     // buffer at max)
     uint8_t buf[256] = {0};
-    uint8_t encoding_header;
-    HARDENED_TRY(little_endian_encode(digest_len_bits, buf, &encoding_header));
-    buf[encoding_header] = encoding_header;
-
-    // Because the MSG_FIFO is also little endian we need to write in reverse
-    // order
-    for (size_t i = 0; i < encoding_header; i++) {
-      abs_mmio_write8(
-          kKmacBaseAddr + KMAC_MSG_FIFO_REG_OFFSET + ((message.len + i) % 4),
-          buf[encoding_header - i - 1]);
+    uint8_t bytes_written;
+    HARDENED_TRY(little_endian_encode(digest_len_bits, buf, &bytes_written));
+    buf[bytes_written] = bytes_written;
+
+    // Because the MSG_FIFO is interpreted as little endian, we need to write
+    // in reverse order.
+    for (size_t i = 0; i < bytes_written; i++) {
+      abs_mmio_write8(kKmacBaseAddr + KMAC_MSG_FIFO_REG_OFFSET,
+                      buf[bytes_written - i - 1]);
     }
 
-    // Finally write `encoding_header` as the last byte
-    abs_mmio_write8(
-        kKmacBaseAddr + KMAC_MSG_FIFO_REG_OFFSET + ((message.len + i) % 4),
-        buf[encoding_header]);
+    // Finally write `bytes_written` as the last byte.
+    abs_mmio_write8(kKmacBaseAddr + KMAC_MSG_FIFO_REG_OFFSET,
```

Comment:
I see in the diff that there is a `((message.len + i) % 4` that is dropped in the address calculation. `MSG_FIFO` register description states that:

> Any write to this window will be appended to the FIFO. Only lower 2 bits [1:0] of the address matter to writes within the window in order to handle with sub-word writes.

so (without looking at the RTL code) it sounds to me like we need this module 4 value here.

Reply:
You're right, according to the documentation this way shouldn't work. The fact that the tests pass anyway is confusing to me... I'll add the modulo 4 back for now, but long-term it would be good to know if the RTL and documentation disagree so we can fix the docs. (I tried to look a the RTL and couldn't find anywhere that seems to be looking at the lower 2 bits, it seems to be only looking at the mask and blindly packing. But my knowledge of RTL is too weak to know for sure; would appreciate help if you have time @ballifatih !)

Reply:
It also took me quite some time to figure out how it works in RTL.

When SW writes a byte, within TL-UL transaction, there is also a 32-bit mask value that tells which bits of the incoming 32-bit word is valid. This sounds almost like a bit of a redundancy on top of the least significant 2 bits of the address.

Inside MSG_FIFO, there is a prim_packer primitive, whose job is to accumulate incoming mixture of byte/half-word/half writes, and then release a word when sufficient bits are collected. While doing this, prim_packer looks at the mask value, counts the number of 1s to understand whether the incoming data is byte/half-word/word. But then, the actual incoming position/alignment is dropped. Instead, prim_packer internally keeps history of received bytes/words to figure out where the position of the incoming byte should be. Hence, from SW side, the least significant two bits are not used at all!

Here is a doc fix: https://github.com/lowRISC/opentitan/pull/19607

Reply:
Thanks, Fatih! This makes sense and thanks for submitting the doc fix so quickly too :slightly_smiling_face: